lib/more_core_extensions/core_ext/process/pause_resume.rb

```diff
@@ -0,0 +1,100 @@
+module MoreCoreExtensions
+  module ProcessPauseResume
+    if File::ALT_SEPARATOR
+      require 'fiddle'
+      ntdll = Fiddle.dlopen('ntdll')
+      kernel32 = Fiddle.dlopen('kernel32')
+
+      NtSuspendProcess = Fiddle::Function.new(
+        ntdll['NtSuspendProcess'],
+        [Fiddle::TYPE_UINTPTR_T],
+        Fiddle::TYPE_INT
+      )
+
+      private_constant :NtSuspendProcess
+
+      NtResumeProcess = Fiddle::Function.new(
+        ntdll['NtResumeProcess'],
+        [Fiddle::TYPE_UINTPTR_T],
+        Fiddle::TYPE_INT
+      )
+
+      private_constant :NtResumeProcess
+
+      OpenProcess = Fiddle::Function.new(
+        kernel32['OpenProcess'],
+        [Fiddle::TYPE_LONG, Fiddle::TYPE_INT, Fiddle::TYPE_LONG],
+        Fiddle::TYPE_UINTPTR_T
+      )
+
+      private_constant :OpenProcess
+
+      CloseHandle = Fiddle::Function.new(
+        kernel32['CloseHandle'],
+        [Fiddle::TYPE_UINTPTR_T],
+        Fiddle::TYPE_INT
+      )
+
+      private_constant :CloseHandle
+
+      PROCESS_SUSPEND_RESUME = 0x00000800
+
+      private_constant :PROCESS_SUSPEND_RESUME
+    end
+
+    # Returns whether or not the given process is running.
+    #
+    def alive?(pid)
+      Process.kill(0, pid)
+      true
+    rescue Errno::ESRCH
+      false
+    end
+
+    # Suspend the process +pid+. If the process isn't running then this is no-op.
+    #
+    def pause(pid)
+      if File::ALT_SEPARATOR
```

Comment:
Maybe use this conditional to determine the how the method is defined, instead of having the conditional in the method everytime it is called?

```ruby
if File::ALT_SEPARATOR
  def pause(pid)
    begin
      handle = OpenProcess.call(PROCESS_SUSPEND_RESUME, 0, pid)

      if handle == 0
        raise SystemCallError, Fiddle.win32_last_error, "OpenProcess"
      end

      if NtSuspendProcess.call(handle) != 0
        raise SystemCallError, Fiddle.win32_last_error, "NtSuspendProcess"
      end
    ensure
      CloseHandle.call(handle) if handle
    end
  end
else
  def pause(pid)
    Process.kill('STOP', pid) if alive?(pid)
  end
end

Reply:
@NickLaMuro ok, updated